src/track/globaltrackcache.cpp

```diff
@@ -30,240 +36,354 @@ TrackRef createTrackRef(const Track& track) {
+    DEBUG_ASSERT(m_strongPtr);
     DEBUG_ASSERT(trackId.isValid());
-    m_trackRef = GlobalTrackCache::instance().updateTrackIdInternal(
-            m_pTrack,
-            m_trackRef,
-            trackId);
-    DEBUG_ASSERT(m_trackRef.getId() == trackId);
-    m_pTrack->initId(trackId);
-    DEBUG_ASSERT(createTrackRef(*m_pTrack) == m_trackRef);
+    if (m_trackRef.getId().isValid()) {
+        // Ignore initializing the same id twice
+        DEBUG_ASSERT(m_trackRef.getId() == trackId);
+    } else {
+        m_trackRef = m_pInstance->initTrackId(
+                m_strongPtr,
+                m_trackRef,
+                trackId);
+    }
+    unlockCache();
+    DEBUG_ASSERT(m_trackRef == createTrackRef(*m_strongPtr));
 }
 
-
 //static
-GlobalTrackCache* volatile GlobalTrackCache::s_pInstance = nullptr;
-
-//static
-void GlobalTrackCache::createInstance(GlobalTrackCacheEvictor* pEvictor) {
-    DEBUG_ASSERT(s_pInstance == nullptr);
-    s_pInstance = new GlobalTrackCache(pEvictor);
+void GlobalTrackCache::createInstance(GlobalTrackCacheDeleter* pDeleter) {
+    DEBUG_ASSERT(!s_pInstance);
+    s_pInstance = new GlobalTrackCache(pDeleter);
 }
 
 //static
 void GlobalTrackCache::destroyInstance() {
-    DEBUG_ASSERT(s_pInstance != nullptr);
+    DEBUG_ASSERT(s_pInstance);
     GlobalTrackCache* pInstance = s_pInstance;
+    // Reset the static/global pointer before entering the destructor
     s_pInstance = nullptr;
+    // Delete the singular instance
     delete pInstance;
 }
 
+void GlobalTrackCache::deleteTrack(Track* plainPtr) {
+    VERIFY_OR_DEBUG_ASSERT(plainPtr) {
+        kLogger.warning()
+                << "Cannot delete null track pointer";
+        return;
+    }
+    if (traceLogEnabled()) {
+        plainPtr->dumpObjectInfo();
+    }
+    // Track object must not be deleted by operator new!
+    // Otherwise the deleted track object might be accessed
+    // when processing cross-thread signals that are delayed
+    // within a queued connection and may arrive after the
+    // object has already been deleted.
+    if (debugLogEnabled()) {
+        kLogger.debug()
+                << "Deleting track"
+                << plainPtr;
+    }
+    plainPtr->deleteLater();
+}
+
 //static
-void GlobalTrackCache::deleter(Track* pTrack) {
-    if (s_pInstance != nullptr) {
-        s_pInstance->evict(pTrack);
+void GlobalTrackCache::deleter(Track* plainPtr) {
+    DEBUG_ASSERT(plainPtr);
+    // Any access to plainPtr is forbidden!! Due to race condition
+    // this pointer might already have been deleted! This happens
+    // when a previous invocation is outpaced by a following
+    // invocation that will already delete the object behind the
+    // pointer. Tracks might be revived before being deleted and
+    // as a consequence the reference count might drop to 0 multiple
+    // times. The order in which those competing invocations finally
+    // lock and enter the thread-safe cache is undefined!!!
```

Comment:
How about this comment instead:    

    // Any access to plainPtr before a validity check inside the
    // GlobalTrackCacheLocker scope is forbidden!!
    // This pointer might already have been by a second shared_ptr instance
    // revived from a plain pointer while we where suspendend here in the
    // deleter call.